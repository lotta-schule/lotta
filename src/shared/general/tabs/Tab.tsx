import * as React from 'react';
import { ButtonProps } from '../button/Button';
import { NavigationButton } from '../button/NavigationButton';
import clsx from 'clsx';

import styles from './Tab.module.scss';

export type TabProps = Omit<ButtonProps, 'onClick' | 'ref'> & {
    value: string | number;
    selected?: boolean;
};

export const Tab: React.FC<TabProps> = ({ className, children, ...props }) => {
    return (
        <NavigationButton
            className={clsx(styles.tab, className)}
            role={'tab'}
            {...props}
        >
            {children}
        </NavigationButton>
    );
};
Tab.displayName = 'Tab';
