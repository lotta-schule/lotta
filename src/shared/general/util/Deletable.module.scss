.root {
    display: inline-block;
    position: relative;
    .button {
        position: absolute;
        top: -7.5px;
        right: -7.5px;
        width: 15px;
        height: 15px;
        overflow: hidden;
        z-index: 1;
        transform: scale(0);
    }
    &:hover {
        .button {
            transform: scale(1);
        }
    }
}
