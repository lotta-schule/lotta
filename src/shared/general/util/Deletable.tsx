import * as React from 'react';
import { Close } from '@material-ui/icons';
import { Button } from '../button/Button';
import clsx from 'clsx';

import styles from './Deletable.module.scss';

export interface DeletableProps
    extends Omit<React.HTMLProps<HTMLDivElement>, 'onClick'> {
    className?: string;
    title?: string;
    onDelete?: React.MouseEventHandler<HTMLButtonElement> | null;
}

export const Deletable: React.FC<DeletableProps> = ({
    children,
    onDelete,
    title = 'lÃ¶schen',
    className,
    ...props
}) => {
    return (
        <div className={clsx(styles.root, className)} {...props}>
            {onDelete && (
                <Button
                    small
                    className={styles.button}
                    icon={<Close />}
                    title={title}
                    onClick={onDelete ?? undefined}
                />
            )}
            {children}
        </div>
    );
};
