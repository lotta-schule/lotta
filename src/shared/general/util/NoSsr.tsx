import * as React from 'react';

export const NoSsr: React.FC = ({ children }) => {
    const [hasRendered, setHasRendered] = React.useState(false);
    React.useEffect(() => {
        setHasRendered(true);
    }, []);

    return hasRendered ? (children as React.ReactElement) : null;
};
