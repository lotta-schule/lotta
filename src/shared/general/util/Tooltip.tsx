import * as React from 'react';
import { Tooltip as ReachTooltip } from '@reach/tooltip';
import clsx from 'clsx';

import styles from './Tooltip.module.scss';

export interface TooltipProps extends React.HTMLAttributes<HTMLDivElement> {
    label: React.ReactNode;
}

export const Tooltip = React.forwardRef(
    (
        { children, label, className, ...props }: TooltipProps,
        ref: React.ForwardedRef<HTMLDivElement>
    ) => {
        return (
            <ReachTooltip
                ref={ref}
                label={label}
                className={clsx(styles.root, className)}
                {...props}
            >
                {children}
            </ReachTooltip>
        );
    }
);
Tooltip.displayName = 'Tooltip';
