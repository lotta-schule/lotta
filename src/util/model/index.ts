export * from './Tenant';
export * from './User';
export * from './Widget';