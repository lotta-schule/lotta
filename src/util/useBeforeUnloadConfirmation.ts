import { useRouter } from 'next/router';
import * as React from 'react';

export const useBeforeUnloadConfirmation = (
    enabled: boolean,
    message: string
) => {
    const lastHistoryState = React.useRef(global.history?.state);
    const router = useRouter();
    React.useEffect(() => {
        const storeLastHistoryState = () => {
            lastHistoryState.current = history.state;
        };
        router.events.on('routeChangeComplete', storeLastHistoryState);
        return () => {
            router.events.off('routeChangeComplete', storeLastHistoryState);
        };
    }, [router]);

    React.useEffect(() => {
        let isWarned = false;

        const routeChangeStart = (url: string) => {
            if (router.asPath !== url && enabled && !isWarned) {
                isWarned = true;
                if (confirm(message)) {
                    router.push(url);
                } else {
                    isWarned = false;
                    router.events.emit('routeChangeError');

                    // HACK
                    const state = lastHistoryState.current;
                    if (
                        state != null &&
                        history.state != null &&
                        state.idx !== history.state.idx
                    ) {
                        history.go(state.idx < history.state.idx ? -1 : 1);
                    }

                    throw 'Abort route change. Please ignore this error.';
                }
            }
        };

        const beforeUnload = (e: Event) => {
            if (enabled && !isWarned) {
                const event = e || window.event;
                event.returnValue = true;
                return message;
            }
            return null;
        };

        router.events.on('routeChangeStart', routeChangeStart);
        window.addEventListener('beforeunload', beforeUnload);

        return () => {
            router.events.off('routeChangeStart', routeChangeStart);
            window.removeEventListener('beforeunload', beforeUnload);
        };
    }, [message, router, enabled]);
};
