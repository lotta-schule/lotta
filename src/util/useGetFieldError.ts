import { useCallback } from 'react';
import { ApolloError } from '@apollo/client';

export const useGetFieldError = (error?: ApolloError) => {
    return useCallback(
        (fieldName: string): string | false => {
            if (error) {
                const messages: string[] | undefined =
                    error.graphQLErrors &&
                    (error.graphQLErrors.map(
                        (graphQLError: any) =>
                            graphQLError &&
                            graphQLError.details &&
                            graphQLError.details[fieldName]
                    ) as string[] | undefined);
                if (messages) {
                    return messages[0];
                }
            }
            return false;
        },
        [error]
    );
};
