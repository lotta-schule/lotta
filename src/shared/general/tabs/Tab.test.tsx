import { render } from 'test/util';
import { Tab } from './Tab';

describe('tabs/Tab', () => {
    it('should render a Tab button', () => {
        const screen = render(<Tab value={4}>Tab-Button</Tab>);
        expect(screen.container).toMatchSnapshot();
    });

    it('should make a tab button selected', () => {
        const screen = render(
            <Tab selected value={4}>
                Tab-Button
            </Tab>
        );
        expect(screen.getByRole('tab', { selected: true })).toBeVisible();
    });
});
