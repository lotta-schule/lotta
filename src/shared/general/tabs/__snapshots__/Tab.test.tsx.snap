// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tabs/Tab should render a Tab button 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <button
      aria-selected="false"
      class="lotta-base-button variant__default lotta-button lotta-navigation-button tab"
      role="tab"
      type="button"
      value="4"
    >
      <span
        class="lotta-button_text"
      >
        Tab-Button
      </span>
    </button>
  </div>
</div>
`;
